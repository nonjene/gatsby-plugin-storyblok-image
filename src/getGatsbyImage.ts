import type { IGatsbyImageData } from 'gatsby-plugin-image/dist/src/components/gatsby-image.browser'

import getBasicImageProps from './utils/getBasicImageProps'
import getWidths from './utils/getWidths'
import buildImageUrl, { buildLowFiUrl } from './utils/buildImageUrl'
import { defaultOptions } from './defaults'
import { GetGatsbyImageOptions, MergedGetGatsbyImageOptions } from './typings/module'

export default function getGatsbyImage(imageRaw: string, args: GetGatsbyImageOptions): IGatsbyImageData | null {
  if (imageRaw.endsWith('.svg')) {
    return null
  }
  // reset when url containing image config
  const image = imageRaw.replace(/\/m\/[^.]*$/, '')

  const imageProps = getBasicImageProps(image)

  // url not identified as Storyblok image
  if (!imageProps) {
    return null
  }

  const options: MergedGetGatsbyImageOptions = {
    ...defaultOptions,
    ...args
  }
  const {
    metadata: { dimensions },
    originalPath
  } = imageProps

  const width = options.width || 0
  const height = options.height || 0

  const desiredAspectRatio = width && height ? width / height : dimensions.aspectRatio

  let outputWidth: number
  let outputHeight: number
  let sizes: string

  // Get output width and height

  switch (true) {
    case options.layout === 'fullWidth':
      outputWidth = desiredAspectRatio
      outputHeight = 1
      break
    case width && !height:
      outputWidth = width
      outputHeight = Math.round(width / desiredAspectRatio)
      break
    case Boolean(!width && height):
      outputWidth = Math.round(height * desiredAspectRatio)
      outputHeight = height
      break
    case Boolean(width && height):
      outputWidth = width
      outputHeight = height
      break
    case !width && !height:
    default:
      outputWidth = dimensions.width
      outputHeight = dimensions.height
      break
  }

  switch (options.layout) {
    case 'constrained':
      sizes = `(min-width: ${outputWidth}px) ${outputWidth}px, 100vw`
      break
    case 'fullWidth':
      sizes = '100vw'
      break
    case 'fixed':
    default:
      sizes = `${outputWidth}px`
      break
  }

  const widths = getWidths(options, outputWidth, dimensions.width)
  const srcSetArr = widths.map((currentWidth) => {
    const resolution = `${currentWidth}w`

    const currentHeight = Math.round(currentWidth / desiredAspectRatio)

    // Workaround for Storyblok has bug when crop with width larger than 4000 (eg: crop 5000 will get 4000)
    // Not to set crop when size is same to source
    const sameToSource = currentWidth === dimensions.width && currentHeight === dimensions.height

    const url = buildImageUrl(originalPath, {
      ...options,
      width: sameToSource ? undefined : currentWidth,
      height: sameToSource ? undefined : currentHeight
    })

    return `${url} ${resolution}`
  })
  const src = buildImageUrl(originalPath, { ...options, width, height })

  return {
    layout: options.layout,
    placeholder: {
      sources: [],
      fallback: options.fallback === false ? null : buildLowFiUrl(originalPath, options)
    },
    backgroundColor: options.backgroundColor || 'transparent',
    images: {
      fallback: {
        src,
        srcSet: srcSetArr.join(',\n') || null,
        sizes
      },
      sources: []
    },
    width: outputWidth,
    height: outputHeight
  }
}
