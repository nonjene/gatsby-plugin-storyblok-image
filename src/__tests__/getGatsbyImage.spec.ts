import getGatsbyImage from '../getGatsbyImage'

describe('getGatsbyImage', () => {
  const IMG_URL = 'https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png'

  describe('when layout is fixed', () => {
    it('should match the return data when provide width & height & quality', () => {
      expect(
        getGatsbyImage(IMG_URL, {
          layout: 'fixed',
          width: 100,
          height: 100,
          quality: 100
        })
      ).toMatchSnapshot()
    })

    it('should match the return data when provide only width', () => {
      expect(
        getGatsbyImage(IMG_URL, {
          layout: 'fixed',
          width: 100
        })
      ).toMatchSnapshot()
    })
    it('should match the return data when provide only height', () => {
      expect(
        getGatsbyImage(IMG_URL, {
          layout: 'fixed',
          height: 100
        })
      ).toMatchSnapshot()
    })

    it('should use original width & height when both not provided', () => {
      expect(
        getGatsbyImage(IMG_URL, {
          layout: 'fixed'
        })
      ).toEqual({
        layout: 'fixed',
        placeholder: {
          sources: [],
          fallback: IMG_URL + '/m/20x0/filters:quality(70)'
        },
        backgroundColor: 'transparent',
        images: {
          sources: [],
          fallback: {
            src: `${IMG_URL}/m/smart`,
            srcSet: `${IMG_URL}/m/smart 1200w`,
            sizes: '1200px'
          }
        },
        width: 1200,
        height: 1000
      })
    })
  })

  describe('when layout is constrained', () => {
    it('should match the return data', () => {
      expect(
        getGatsbyImage(IMG_URL, {
          layout: 'constrained',
          width: 100,
          height: 100
        })
      ).toMatchSnapshot()
    })
    it('should keep aspect ratio when width is larger than the source', () => {
      expect(
        getGatsbyImage(IMG_URL, {
          layout: 'constrained',
          width: 1300,
          height: 1000
        })?.images.fallback.srcSet
      ).toMatchInlineSnapshot(`
        "https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/750x577/smart 750w,
        https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/1080x831/smart 1080w,
        https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/1200x923/smart 1200w"
      `)
    })
  })

  describe('when layout is full width', () => {
    it('should match the return data', () => {
      expect(
        getGatsbyImage(IMG_URL, {
          layout: 'fullWidth'
        })
      ).toMatchSnapshot()
    })
  })
})
