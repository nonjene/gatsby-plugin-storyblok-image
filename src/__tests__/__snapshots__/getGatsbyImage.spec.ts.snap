// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getGatsbyImage when layout is constrained should match the return data 1`] = `
Object {
  "backgroundColor": "transparent",
  "height": 100,
  "images": Object {
    "fallback": Object {
      "sizes": "(min-width: 100px) 100px, 100vw",
      "src": "https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/100x100/smart/filters:quality(100)",
      "srcSet": "https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/100x100/smart/filters:quality(100) 100w,
https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/200x200/smart/filters:quality(100) 200w,
https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/300x300/smart/filters:quality(100) 300w,
https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/750x750/smart/filters:quality(100) 750w,
https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/1080x1080/smart/filters:quality(100) 1080w,
https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/1200x1200/smart/filters:quality(100) 1200w",
    },
    "sources": Array [],
  },
  "layout": "constrained",
  "placeholder": Object {
    "fallback": "https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/20x0/filters:quality(10)",
    "sources": Array [],
  },
  "width": 100,
}
`;

exports[`getGatsbyImage when layout is fixed should match the return data when provide only height 1`] = `
Object {
  "backgroundColor": "transparent",
  "height": 100,
  "images": Object {
    "fallback": Object {
      "sizes": "120px",
      "src": "https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/0x100/smart/filters:quality(100)",
      "srcSet": "https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/120x100/smart/filters:quality(100) 120w,
https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/240x200/smart/filters:quality(100) 240w,
https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/360x300/smart/filters:quality(100) 360w",
    },
    "sources": Array [],
  },
  "layout": "fixed",
  "placeholder": Object {
    "fallback": "https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/20x0/filters:quality(10)",
    "sources": Array [],
  },
  "width": 120,
}
`;

exports[`getGatsbyImage when layout is fixed should match the return data when provide only width 1`] = `
Object {
  "backgroundColor": "transparent",
  "height": 83,
  "images": Object {
    "fallback": Object {
      "sizes": "100px",
      "src": "https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/100x0/smart/filters:quality(100)",
      "srcSet": "https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/100x83/smart/filters:quality(100) 100w,
https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/200x167/smart/filters:quality(100) 200w,
https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/300x250/smart/filters:quality(100) 300w",
    },
    "sources": Array [],
  },
  "layout": "fixed",
  "placeholder": Object {
    "fallback": "https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/20x0/filters:quality(10)",
    "sources": Array [],
  },
  "width": 100,
}
`;

exports[`getGatsbyImage when layout is fixed should match the return data when provide width & height 1`] = `
Object {
  "backgroundColor": "transparent",
  "height": 100,
  "images": Object {
    "fallback": Object {
      "sizes": "100px",
      "src": "https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/100x100/smart/filters:quality(100)",
      "srcSet": "https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/100x100/smart/filters:quality(100) 100w,
https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/200x200/smart/filters:quality(100) 200w,
https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/300x300/smart/filters:quality(100) 300w",
    },
    "sources": Array [],
  },
  "layout": "fixed",
  "placeholder": Object {
    "fallback": "https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/20x0/filters:quality(10)",
    "sources": Array [],
  },
  "width": 100,
}
`;

exports[`getGatsbyImage when layout is full width should match the return data 1`] = `
Object {
  "backgroundColor": "transparent",
  "height": 1,
  "images": Object {
    "fallback": Object {
      "sizes": "100vw",
      "src": "https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/smart/filters:quality(100)",
      "srcSet": "https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/750x625/smart/filters:quality(100) 750w,
https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/1080x900/smart/filters:quality(100) 1080w,
https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/1200x1000/smart/filters:quality(100) 1200w",
    },
    "sources": Array [],
  },
  "layout": "fullWidth",
  "placeholder": Object {
    "fallback": "https://a.storyblok.com/f/143948/1200x1000/99335d0004/bk-1200-hoang-lan-ph-m.png/m/20x0/filters:quality(10)",
    "sources": Array [],
  },
  "width": 1.2,
}
`;
